Tests/ContainerizationExtrasTests/TestAsyncLock.swift

```diff
@@ -0,0 +1,420 @@
+            
+            func append(_ result: String) {
+                results.append(result)
+            }
+            
+            func getResults() -> [String] {
+                return results
+            }
+            
+            func count() -> Int {
+                return results.count
+            }
+        }
+        
+        let store = ResultStore()
+        
+        await withTaskGroup(of: Void.self) { group in
+            for i in 0..<5 {
+                group.addTask {
+                    await lock.withLock { context in
+                        // Simulate async work
+                        try? await Task.sleep(nanoseconds: 10_000_000) // 10ms
+                        await store.append("task-\(i)")
+                    }
+                }
+            }
+            
+            await group.waitForAll()
+        }
+        
+        let count = await store.count()
+        #expect(count == 5)
+        
+        // Results should all be unique (no concurrent modification)
+        let results = await store.getResults()
+        #expect(Set(results).count == 5)
+    }
+    
+    @Test func testLockWithThrowingOperation() async throws {
+        let lock = AsyncLock()
+        
+        actor Counter {
+            private var value = 0
+            
+            func increment() -> Int {
+                value += 1
+                return value
+            }
+            
+            func getValue() -> Int {
+                return value
+            }
+        }
+        
+        let counter = Counter()
+        
+        struct TestError: Error {}
+        
+        do {
+            try await lock.withLock { context in
+                let _ = await counter.increment()
+                throw TestError()
+            }
+            #expect(Bool(false), "Should have thrown an error")
+        } catch is TestError {
+            // Expected error
+        }
+        
+        // Lock should still work after an error
+        let _ = await lock.withLock { context in
+            await counter.increment()
+        }
+        
+        let finalCount = await counter.getValue()
+        #expect(finalCount == 2)
+    }
+    
+    @Test func testLockReentrancyPrevention() async throws {
+        let lock = AsyncLock()
+        
+        actor ExecutionTracker {
+            private var order: [String] = []
+            
+            func append(_ event: String) {
+                order.append(event)
+            }
+            
+            func getOrder() -> [String] {
+                return order
+            }
+        }
+        
+        let tracker = ExecutionTracker()
+        
+        await withTaskGroup(of: Void.self) { group in
+            // First task - holds lock for a while
+            group.addTask {
+                await lock.withLock { context in
+                    await tracker.append("task1-start")
+                    try? await Task.sleep(nanoseconds: 50_000_000) // 50ms
```

Comment:
Task.sleep has a for: constructor that takes in enums for different time intervals and should look much nicer:

try await Task.sleep(for: .milliseconds(50))